### PR TITLE
Explain why we require `_` for empty patterns

### PR DESCRIPTION
**Pull request by Nadrieril**
This adds a note to the "non-exhaustive patterns" diagnostic to explain why we sometimes require extra `_` patterns on empty types. This is one of the two diagnostic improvements I wanted to do before [stabilizing `min_exhaustive_patterns`](https://github.com/rust-lang/rust/pull/122792).

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 1]
```diff
@@ -998,27 +998,31 @@ fn report_non_exhaustive_match<'p, 'tcx>(
...
+                } else if ty.inner() == cx.tcx.types.isize {
                     err.note(format!(
                         "`{ty}` does not have fixed minimum and maximum values, so half-open ranges are necessary to match \
                              exhaustively",
                     ));
                 }
-            } else if ty == cx.tcx.types.str_ {
+            } else if ty.inner() == cx.tcx.types.str_ {
                 err.note("`&str` cannot be matched exhaustively, so a wildcard `_` is necessary");
-            } else if cx.is_foreign_non_exhaustive_enum(cx.reveal_opaque_ty(ty)) {
```
Comment (compiler-errors) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:1020]: Where did the reveal go?

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 2]
```diff
@@ -1168,22 +1172,22 @@ fn joined_uncovered_patterns<'p, 'tcx>(
     }
 }
 
-fn collect_non_exhaustive_tys<'tcx>(
+/// Collect types that require specific explanations when they show up in witnesses.
+fn collect_special_tys<'tcx>(
     cx: &PatCtxt<'_, 'tcx>,
     pat: &WitnessPat<'_, 'tcx>,
-    non_exhaustive_tys: &mut FxIndexSet<Ty<'tcx>>,
+    special_tys: &mut FxIndexSet<RevealedTy<'tcx>>,
```
Comment (compiler-errors) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:1179]: oh lol nvm, it was already revealed it seems?

**Review by compiler-errors — APPROVED** [review]

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_mir_build/src/thir/pattern/check_match.rs:1179]: i.e. we were peeling off the fact that we knew it was revealed previously

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/check_match.rs:1179]: yep exactly

**Pull request merged**